src/runner/mod.rs

```diff
@@ -0,0 +1,190 @@
+    pub fn run<T>(self, suite: (Suite<T>, T)) -> SuiteReport
+    where
+        T: Clone + Send + Sync + ::std::fmt::Debug,
+    {
+        let (suite, mut environment) = suite;
+        self.prepare_before_run();
+        let report = self.visit(&suite, &mut environment);
+        self.clean_after_run();
+        report
+    }
+
+    pub fn run_or_exit<T>(self, suite: (Suite<T>, T))
+    where
+        T: Clone + Send + Sync + ::std::fmt::Debug,
+    {
+        if self.run(suite).failed > 0 {
+            // XXX Cargo test failure returns 101.
+            //
+            // > "We use 101 as the standard failure exit code because it's something unique
+            // > that the test runner can check for in run-fail tests (as opposed to something
+            // > like 1, which everybody uses). I don't expect this behavior can ever change.
+            // > This behavior probably dates to before 2013,
+            // > all the way back to the creation of compiletest." – @brson
+
+            ::std::process::exit(101);
+        }
+    }
+
+    fn broadcast(&self, event: Event) {
+        for mutex in &self.handlers {
+            if let Ok(ref mut handler) = mutex.lock() {
+                handler.handle(&event);
+            } else {
+                println!("Error: lock failed");
+            }
+        }
+    }
+
+    fn prepare_before_run(&self) {
+        panic::set_hook(Box::new(|_panic_info| {
+            // XXX panics already catched at the test call site, don't output the trace in stdout
+        }));
+    }
+
+    fn clean_after_run(&self) {
+        // XXX reset panic hook back to default hook:
+        let _ = panic::take_hook();
+    }
+}
+
+impl<T> Visitor<Suite<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, suite: &Suite<T>, environment: &mut Self::Environment) -> Self::Output {
+        self.broadcast(Event::EnterSuite(suite.info.clone()));
+        let report = self.visit(&suite.context, environment);
+        self.broadcast(Event::ExitSuite(report.clone()));
+        report
+    }
+}
+
+impl<T> Visitor<ContextMember<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, member: &ContextMember<T>, environment: &mut Self::Environment) -> Self::Output {
+        match member {
+            &ContextMember::Example(ref example) => {
+                self.visit(example, environment).into()
+            }
+            &ContextMember::Context(ref context) => {
+                self.visit(context, &mut environment.clone())
+            }
+        }
+    }
+}
+
+impl<T> Visitor<Context<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, context: &Context<T>, environment: &mut Self::Environment) -> Self::Output {
```

Comment:
Yeah. I've been trying to split things up, but most attempts turned out to be quite involved and often less readable. I'll keep digging though.